spec/replication_spec.cr

```diff
@@ -65,3 +65,133 @@ describe LavinMQ::Replication::Client do
     end
   end
 end
+
+describe LavinMQ::Replication::Server do
+  data_dir = "/tmp/lavinmq-follower"
+
+  before_each do
+    FileUtils.rm_rf data_dir
+    Dir.mkdir_p data_dir
+    File.write File.join(data_dir, ".replication_secret"), Server.@replicator.@password, 0o400
+    Server.vhosts["/"].declare_queue("repli", true, false)
+  end
+
+  after_each do
+    FileUtils.rm_rf data_dir
+  end
+
+  it "should shut down gracefully" do
+    repli = LavinMQ::Replication::Client.new(data_dir)
+    3.times do
+      spawn do
+        repli.follow("127.0.0.1", LavinMQ::Config.instance.replication_port)
+      end
+    end
+  end
+
+  describe "min_followers" do
+    before_each do
+      LavinMQ::Config.instance.min_followers = 1
+    end
+
+    after_each do
+      LavinMQ::Config.instance.min_followers = 0
+    end
+    it "should publish when min_followers is fulfilled" do
+      q = Server.vhosts["/"].queues["repli"].as(LavinMQ::Queue)
+      repli = LavinMQ::Replication::Client.new(data_dir)
+      spawn do
+        repli.follow("127.0.0.1", LavinMQ::Config.instance.replication_port)
+      end
+      with_channel do |ch|
+        ch.basic_publish "hello world", "", "repli"
+      end
+      q.basic_get(true) { }.should be_true
+      repli.close
+    end
+
+    it "should not publish when min_followers is not fulfilled" do
+      done = Channel(Nil).new
+      client : AMQP::Client::Connection? = nil
+      spawn do
+        with_channel do |ch, conn|
+          client = conn
+          q = ch.queue("repli")
+          q.publish_confirm "hello world"
+          done.send nil
+        end
+      end
+      select
+      when done.receive
+        fail "Should not receive message"
+      when timeout(0.1.seconds)
+        #ugly hack to release replicator from waiting for lag
```

Comment:
Had to work around `spec_helpsers` `after_each()` that calls `server.stop` here.. because `server.stop` only calls `replicator.clear`, which will not finish if we are waiting for lag.